# Dataset

We will start exploring the parameters of the dataset to learn what data is available.

In [1]:
import pandas as pd

First we need to access the dataset. We will use the AllenSDK and the BrainObservatoryCache to do so. First we need to set this up - the key step is to provide a <b>manifest file</b>. The SDK uses this file to know what data is available and organize the files it downloads. If you instantiate the BrainObservatoryCache without proviing a manifest file, it will create one in your working directory.

In [2]:
from allensdk.core.brain_observatory_cache import BrainObservatoryCache
manifest_file = r'/Users/saskiad/Code/brain_observatory/manifest_v1point3.json'
boc = BrainObservatoryCache(manifest_file=manifest_file)

We can use the BrainObservatoryCache to explore the parameters of the dataset. 

(section-label)=
## Targeted structures
What brain regions were recorded across the dataset? To determine this we use a function called <b>get_all_targeted_structures</b> to create a list of the regions.

In [3]:
boc.get_all_targeted_structures()

['VISal', 'VISam', 'VISl', 'VISp', 'VISpm', 'VISrl']

We see that data was collected in six different visual areas. VISp is the primary visual cortex, also known as V1. The others are higher visual areas that surround VISp. See [visual cortical areas](anatomy.md) to learn more about these areas and how we map them.

(section-label)=
## Cre lines and reporters
We used Cre lines to drive the expression of GCaMP6 in specific populations of neurons. We can find a list of all the cre lines used in this dataset with a similar function

In [4]:
boc.get_all_cre_lines()

['Cux2-CreERT2',
 'Emx1-IRES-Cre',
 'Fezf2-CreER',
 'Nr5a1-Cre',
 'Ntsr1-Cre_GN220',
 'Pvalb-IRES-Cre',
 'Rbp4-Cre_KL100',
 'Rorb-IRES2-Cre',
 'Scnn1a-Tg3-Cre',
 'Slc17a7-IRES2-Cre',
 'Sst-IRES-Cre',
 'Tlx3-Cre_PL56',
 'Vip-IRES-Cre']

Cre is a driver that drives the expression of a reporter. We used four different reporter lines in this dataset.

In [5]:
boc.get_all_reporter_lines()

['Ai148(TIT2L-GC6f-ICL-tTA2)',
 'Ai162(TIT2L-GC6s-ICL-tTA2)',
 'Ai93(TITL-GCaMP6f)',
 'Ai93(TITL-GCaMP6f)-hyg',
 'Ai94(TITL-GCaMP6s)']

```{note}
Reporter lines: All the experiments in this dataset use GCaMP6. The large majority use GCaMP6f and only a few use GCaMP6s. However, you see four different reporters listed here. Why is this? Ai93 is the GCaMP6f reporter we used with the excitatory Cre lines. However, this reporter does not work well for inhibitory Cre lines. We used Ai148, another GCaMP6f reporter, with Vip-IRES-Cre and Sst-IRES-Cre. However, this didn't work with the Pvalb-IRES-Cre. We use Ai162, a GCaMP6s reporter with Pvalb. Additionally, to have a GCaMP6f vs GCaMP6s comparison, we collected a small number of experiments using Ai94 with the Slc17a7-IRES2-Cre. This is a GCaMP6s reporter that complements Ai93. Slc17a7-IRES2-Cre is the only Cre line that was recorded using multiple reporter types. 
```

See [Transgenic tools](transgenic-tools.md) to learn more about these Cre lines and reporters.

(section-label)=
## Imaging depths
Each experiment was collected at a single imaging depth.

In [6]:
boc.get_all_imaging_depths()

[175,
 185,
 195,
 200,
 205,
 225,
 250,
 265,
 275,
 276,
 285,
 300,
 320,
 325,
 335,
 350,
 365,
 375,
 390,
 400,
 550,
 570,
 625]

These values are in um below the surface of the cortex. This is a long list and some of the values don't differ by very much. How meaningful is it? We roughly consider depths less than 250 to be layer 2/3, 250-350 to be layer 4, 350-500 to be layer 5, and over 500 to be layer 6. Keep in mind, much of the imaging here was done with layer specific Cre lines, so for most purposes the best way to get layer specificity is to select appropriate Cre lines.

(section-label)=
## Visual stimuli
What were the visual stimuli that we showed to the mice?

In [7]:
boc.get_all_stimuli()

['drifting_gratings',
 'locally_sparse_noise',
 'locally_sparse_noise_4deg',
 'locally_sparse_noise_8deg',
 'natural_movie_one',
 'natural_movie_three',
 'natural_movie_two',
 'natural_scenes',
 'spontaneous',
 'static_gratings']

MORE ON THESE

(section-label)=
## Experiment containers & sessions

The <b>experiment container</b> describes a set of 3 imaging <b>sessions</b> performed for the same field of view (ie. same targeted structure and imaging depth in the same mouse that targets the same set of neurons). Each experiment container has a unique ID number.

We will identify all the experiment containers for a given stucture and Cre line:

In [8]:
visual_area = 'VISp'
cre_line ='Cux2-CreERT2'

exps = boc.get_experiment_containers(targeted_structures=[visual_area], cre_lines=[cre_line])

```{note}
<b>get_experiment_containers</b> returns all experiment containers that meet the conditions we have specified. If we don't pass any parameters, it returns all experiment containers.
```

We can make a dataframe of the list of experiment containers to see what information we get about them:

In [9]:
pd.DataFrame(exps)

id  imaging_depth targeted_structure      cre_line  \
0   511510736            175               VISp  Cux2-CreERT2   
1   511510855            175               VISp  Cux2-CreERT2   
2   511509529            175               VISp  Cux2-CreERT2   
3   511507650            175               VISp  Cux2-CreERT2   
4   511510699            275               VISp  Cux2-CreERT2   
5   702934962            275               VISp  Cux2-CreERT2   
6   645413757            275               VISp  Cux2-CreERT2   
7   659767480            275               VISp  Cux2-CreERT2   
8   511510650            175               VISp  Cux2-CreERT2   
9   712178509            275               VISp  Cux2-CreERT2   
10  511510779            275               VISp  Cux2-CreERT2   
11  511510667            275               VISp  Cux2-CreERT2   
12  524691282            275               VISp  Cux2-CreERT2   
13  511510670            175               VISp  Cux2-CreERT2   
14  701412138            175               VISp  Cux2-CreERT2   
15  511510718            175               VISp  Cux2-CreERT2   

         reporter_line donor_name                        specimen_name tags  \
0   Ai93(TITL-GCaMP6f)     222426  Cux2-CreERT2;Camk2a-tTA;Ai93-222426   []   
1   Ai93(TITL-GCaMP6f)     229106  Cux2-CreERT2;Camk2a-tTA;Ai93-229106   []   
2   Ai93(TITL-GCaMP6f)     222420  Cux2-CreERT2;Camk2a-tTA;Ai93-222420   []   
3   Ai93(TITL-GCaMP6f)     222424  Cux2-CreERT2;Camk2a-tTA;Ai93-222424   []   
4   Ai93(TITL-GCaMP6f)     225037  Cux2-CreERT2;Camk2a-tTA;Ai93-225037   []   
5   Ai93(TITL-GCaMP6f)     382421  Cux2-CreERT2;Camk2a-tTA;Ai93-382421   []   
6   Ai93(TITL-GCaMP6f)     348262  Cux2-CreERT2;Camk2a-tTA;Ai93-348262   []   
7   Ai93(TITL-GCaMP6f)     360565  Cux2-CreERT2;Camk2a-tTA;Ai93-360565   []   
8   Ai93(TITL-GCaMP6f)     222425  Cux2-CreERT2;Camk2a-tTA;Ai93-222425   []   
9   Ai93(TITL-GCaMP6f)     390323  Cux2-CreERT2;Camk2a-tTA;Ai93-390323   []   
10  Ai93(TITL-GCaMP6f)     225036  Cux2-CreERT2;Camk2a-tTA;Ai93-225036   []   
11  Ai93(TITL-GCaMP6f)     222420  Cux2-CreERT2;Camk2a-tTA;Ai93-222420   []   
12  Ai93(TITL-GCaMP6f)     243293  Cux2-CreERT2;Camk2a-tTA;Ai93-243293   []   
13  Ai93(TITL-GCaMP6f)     225037  Cux2-CreERT2;Camk2a-tTA;Ai93-225037   []   
14  Ai93(TITL-GCaMP6f)     382421  Cux2-CreERT2;Camk2a-tTA;Ai93-382421   []   
15  Ai93(TITL-GCaMP6f)     231584  Cux2-CreERT2;Camk2a-tTA;Ai93-231584   []   

    failed  
0    False  
1    False  
2    False  
3    False  
4    False  
5    False  
6    False  
7    False  
8    False  
9    False  
10   False  
11   False  
12   False  
13   False  
14   False  
15   False

(section-label)=
## Session types
The responses to this full set of visual stimuli were recorded across three imaging sessions. We returned to the same targeted structure and same imaging depth in the same mouse to recorded the same group of neurons across three different days.

In [10]:
boc.get_all_session_types()

['three_session_A', 'three_session_B', 'three_session_C', 'three_session_C2']

We will explore which stimuli are part of which session in REFERENCE. 
During data processing, we matched identified ROIs (REFERENCE) across each of the sessions. Approximately one third of the neurons in the dataset were matched across all three sessions, one third were matched in two of the three session, and one third were only found in one session. Neurons have unique ids, called specimen_ids, that are shared across the sessions they are found in.